File: algorithms/04_java/numerical/EulerMethod.java
import java.util.function.DoubleBinaryOperator;

// Forward Euler method for ODEs.
public class EulerMethod {
    static double eulerMethod(DoubleBinaryOperator f, double y, double t, double t1, int steps) {
        double h = (t1 - t) / steps;
        for (int i = 0; i < steps; i++) { y += h * f.applyAsDouble(t, y); t += h; }
        return y;
    }

    public static void main(String[] args) {
        assert Math.abs(eulerMethod((t, y) -> y, 1, 0, 1, 1000) - Math.E) < 0.01;
        System.out.println("euler method ok");
    }
}

File: algorithms/02_c/numerical/euler_method.c
#include <assert.h>
#include <math.h>
#include <stdio.h>

static double f(double t, double y) { (void)t; return y; }

static double euler_method(double (*func)(double, double), double y, double t, double t1, int steps) {
    double h = (t1 - t) / steps;
    for (int i = 0; i < steps; ++i) {
        y += h * func(t, y);
        t += h;
    }
    return y;
}

int main(void) {
    double value = euler_method(f, 1.0, 0.0, 1.0, 1000);
    assert(fabs(value - M_E) < 0.01);
    printf("euler method ok\n");
    return 0;
}
